cpp/src/structure/select_random_vertices_impl.hpp

```diff
@@ -68,22 +100,52 @@ rmm::device_uvector<vertex_t> select_random_vertices(
     // FIXME: need to double check uniform_random_fill generates random numbers in [0, V) (not [0,
     // V])
     mg_sample_buffer.resize(this_gpu_select_count, handle.get_stream());
-    cugraph::detail::uniform_random_fill(handle.get_stream(),
-                                         mg_sample_buffer.data(),
-                                         mg_sample_buffer.size(),
-                                         vertex_t{0},
-                                         graph_view.number_of_vertices(),
-                                         rng_state);
-  } else {
-    auto local_vertex_partition_range_first = graph_view.local_vertex_partition_range_first();
-    auto local_vertex_partition_range_last  = graph_view.local_vertex_partition_range_last();
 
-    mg_sample_buffer = rmm::device_uvector<vertex_t>(
-      local_vertex_partition_range_last - local_vertex_partition_range_first, handle.get_stream());
-    thrust::sequence(handle.get_thrust_policy(),
+    cugraph::detail::uniform_random_fill(
+      handle.get_stream(),
+      mg_sample_buffer.data(),
+      mg_sample_buffer.size(),
+      vertex_t{0},
+      given_set ? static_cast<vertex_t>((*given_set).size()) : graph_view.number_of_vertices(),
+      rng_state);
+
+    if (given_set) {
+      thrust::gather(handle.get_thrust_policy(),
                      mg_sample_buffer.begin(),
                      mg_sample_buffer.end(),
-                     local_vertex_partition_range_first);
+                     (*given_set).begin(),
+                     mg_sample_buffer.begin());
+    }
+
+  } else {
+    size_t mg_buffer_size = 0;
+    vertex_t start_idx;
+
+    if (given_set) {
+      start_idx = vertex_t{0};
+      if constexpr (multi_gpu) {
+        auto const comm_rank = handle.get_comms().get_rank();
+        auto const comm_size = handle.get_comms().get_size();
+
+        auto recvcounts =
+          host_scalar_allgather(handle.get_comms(), (*given_set).size(), handle.get_stream());
+        std::vector<size_t> displacements(recvcounts.size(), size_t{0});
+        std::partial_sum(recvcounts.begin(), recvcounts.end() - 1, displacements.begin() + 1);
+
+        start_idx = displacements[comm_rank];
```

Comment:
You can call

https://github.com/rapidsai/cugraph/pull/3455/files#diff-7440460f437e37513b76f74237fb74dd00d85fcc02dd5bc359e4048134d1c30fR100
and
https://github.com/rapidsai/cugraph/pull/3455/files#diff-7440460f437e37513b76f74237fb74dd00d85fcc02dd5bc359e4048134d1c30fR78

to better match the `given_set.has_value() == false` case.

Reply:
@seunghwak  You mean given_set.has_value() == true? 
Updated to use `cugraph::partition_manager::compute_partition_range_lasts( ... )`, but didn't understand why we need 
`compute_vertex_partition_id_from_graph_subcomm_ranks`

Reply:
If `static constexpr bool map_major_comm_to_gpu_row_comm` is set to false, you can no longer assume vertex_partition_id == comm_rank.

If you get `vertex_partition_range_lasts = cugraph::partition_manager::compute_partition_range_lasts()`, you need to index `vertex_partition_range_lasts` with vertex partition ID (not comm_rank).